### ISSUE TITLE
[Bots] Picklist/Multiple Checkboxes for true matching does not work

### ISSUE THREAD

**Issue by iammattmartin** [jstanden/cerb#340]
We've tested using picklist for bots to match certain options in that picklist on a group (via custom fields).

When doing so, they are never matched.  Using check boxes (on/off), work perfectly.

**Comment by iammattmartin** [comment]
This also seems to apply to multiple checkboxes.

Normal single checkboxes work fine.

**Comment by iammattmartin** [comment]
Here is an example of what we're seeing.

The tree:
![m6bge](https://user-images.githubusercontent.com/15971590/29251286-21c8dde8-804a-11e7-845a-50b69c6fc7b3.png)

The Yes match (which it skips):
![screen shot 2017-08-13 at 5 09 44 pm](https://user-images.githubusercontent.com/15971590/29251298-41ab33fe-804a-11e7-8934-3870a2423f4f.png)

The data it checks on the message:
![screen shot 2017-08-13 at 5 10 33 pm](https://user-images.githubusercontent.com/15971590/29251307-61151e9e-804a-11e7-85ad-452be910156d.png)

**Comment by jstanden** [comment]
@mattmartincouk In my tests, these custom fields seem to match properly (picklists, multi-checkboxes).

I'm wondering if there's an ordering issue between where you're setting those custom fields and where you're reading them.

I can see from the above examples that this a 'New message added to ticket' event.

Where/when is the 'source' custom field being set?

**Comment by iammattmartin** [comment]
Sorry @jstanden, missed your reply!

We set those in "Before new incoming message is accepted" and the check under "New message added to ticket" so the data should already be there as before would always (or should) be before the new message is added?

**Comment by iammattmartin** [comment]
Also have done more testing with "Before new incoming message is accepted" and referencing a "parent ticket" for our brands, which use a check box for "is any of" X, Y or Z.

Same behaviour in that it does not match as expected.

**Comment by iammattmartin** [comment]
Just for complete thoroughness.

![screen shot 2017-10-12 at 11 19 37 pm](https://user-images.githubusercontent.com/15971590/31522052-ddb35a96-afa3-11e7-8a9d-f5bff1a42024.png)

This is run on before message is accepted. My contact has VIP status so I first tested with my ticket set to unmanaged. It should have matched and followed due to the VIP status, it did not.

Only when I set managed on the ticket did it then do as expected. It completely disregards the checkboxes.

On the simulator, something strange happens.

Set to VIP, not managed - matches but then goes to the next No step.

![screen shot 2017-10-12 at 11 24 07 pm](https://user-images.githubusercontent.com/15971590/31522253-cfe74f02-afa4-11e7-8c20-88dc87948485.png)

Set to VIP and managed - matches and stops as expected, solely based on the managed.

![screen shot 2017-10-12 at 11 24 55 pm](https://user-images.githubusercontent.com/15971590/31522204-947b6eee-afa4-11e7-9b27-e952a98a185f.png)

Set to non-VIP and managed - matches and stops as expected, solely based on the managed.

![screen shot 2017-10-12 at 11 25 33 pm](https://user-images.githubusercontent.com/15971590/31522230-a9f41e4c-afa4-11e7-86a3-a2d4df9c26ac.png)

Hope that helps!

**Comment by jstanden** [comment]
@mattmartincouk So it looks like a potential commonality here is the custom fields being set in the **Before new incoming message is accepted** event, and the new values then **not** being immediately available in subsequent behaviors (whether those are the same behavior, or something like **New message added to a ticket**).

They values are being set in the database, and the conditions work if you update the ticket manually and run the behaviors.

So it seems like an issue with not properly setting those values in the memory model during the behavior's runtime.

I'll take a closer look at the code, but that sounds very plausible.

**Comment by iammattmartin** [comment]
I'm not entirely sure if they work when I update the ticket manually.

I set my contact to VIP which was always ignored both via the simulator and also by sending a real email. No matter where I have used checkboxes, I can never get them to match no matter what.

I had the issue with picklists, so switched to checkboxes although the behaviour was the same and in the debug, it would indicate a non-match as if it couldn't see the content - stuff that had been set a long time before (from memory (my memory)).

To test the theory, is there a way to save checkboxes into the db, reload the memory (/update?) and then test again.

**Comment by jstanden** [comment]
In the code, the **Before new incoming message is accepted** events aren't setting the values in the model within the behavior.  I'm sure that issue carries over into setting it during filtering and checking it on a custom ticket behavior.

In a simpler test, if you set a custom field and then checked it in the same filtering behavior, the value still doesn't exist:

<img width="1119" alt="screen shot 2017-10-12 at 6 34 59 pm" src="https://user-images.githubusercontent.com/63288/31526501-153dcf50-af7c-11e7-9c71-2e9b8da42676.png">

**Comment by jstanden** [comment]
After I sort this part out, I'll test by manually setting the picklist/multi-checkbox values on the contact too.  I just use a separate browser tab and pull up the record in Search->Contacts.

You could add the custom field values to the DB, but it's a little tedious in `custom_field_stringvalue`. You don't have to clear the cache on those since we don't server-side cache the field values between requests.

**Comment by iammattmartin** [comment]
I'll wait before I do any more testing in that case :)

**Comment by jstanden** [comment]
Phew! Found the comparison bug for multi-checkboxes.  Fixed in 8.1.5.

I also found an issue with setting non-ticket custom fields from the filtering event, and fixed that too.

There's still an issue with setting nested custom fields on 'parent ticket' before that ticket exists (i.e. if the incoming message is opening a new ticket).  Something like `parent_ticket_initial_sender_contact_` won't exist during the behavior.  We can sort that out later.  Fields on `parent_ticket_` itself does work, as does `sender_`.

I'll push to GitHub shortly.
